### PR TITLE
NH : New api version to make SAS keys optional

### PR DESCRIPTION
**Pull request by smithaborkar**
New api version to make SAS keys optional
1. Creation of authRules will only take Access Rights
2. If we need to update the keys, RegenerateKeys is a new api exposed
3. Keys can be viewed only through ListKeys Apis

### CODE REVIEW

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 1]
```diff
@@ -1660,33 +1654,33 @@
       "properties": {
         "id": {
           "type": "string",
-          "description": "Gets or sets the id of the created NotificatioHub."
+          "description": "Gets or sets the id of the created NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: You should just say "The id of the created NotificationHub ." -- individual language generators in AutorRest will add the Gets or Sets if it's appropriate (some languages don't support properties)

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 2]
```diff
@@ -1660,33 +1654,33 @@
...
         },
         "location": {
           "type": "string",
-          "description": "Gets or sets datacenter location of the NotificatioHub."
+          "description": "Gets or sets datacenter location of the NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 3]
```diff
@@ -1660,33 +1654,33 @@
...
         },
         "name": {
           "type": "string",
-          "description": "Gets or sets name of the NotificatioHub."
+          "description": "Gets or sets name of the NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 4]
```diff
@@ -1660,33 +1654,33 @@
...
         },
         "type": {
           "type": "string",
-          "description": "Gets or sets resource type of the NotificatioHub."
+          "description": "Gets or sets resource type of the NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 5]
```diff
@@ -1660,33 +1654,33 @@
...
         },
         "tags": {
           "type": "object",
           "additionalProperties": {
             "type": "string"
           },
-          "description": "Gets or sets tags of the NotificatioHub."
+          "description": "Gets or sets tags of the NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 6]
```diff
@@ -1660,33 +1654,33 @@
...
         },
         "properties": {
           "$ref": "#/definitions/NotificationHubProperties",
-          "description": "Gets or sets properties of the NotificatioHub."
+          "description": "Gets or sets properties of the NotificationHub."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2014-09-01/swagger/notificationhubs.json [hunk 7]
```diff
@@ -1660,33 +1654,33 @@
...
         }
       },
-      "description": "Description of a NotificatioHub Resource."
+      "description": "Description of a NotificationHub Resource."
```
Comment (John-Hart) [inline arm-notificationhubs/2014-09-01/swagger/notificationhubs.json]: Same as above

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 8]
```diff
@@ -0,0 +1,1742 @@
...
+        }
+    },
+    "paths": {
+        "/subscriptions/{subscriptionId}/providers/Microsoft.NotificationHubs/checkNamespaceAvailability": {
+            "post": {
+                "tags": [
+                    "Namespaces"
+                ],
+                "operationId": "Namespaces_CheckAvailability",
+                "description": "Checks the availability of the given service namespace across all Windows Azure subscriptions. This is useful because the domain name is created based on the service namespace name.",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Windows Azure -> only "Azure"

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 9]
```diff
@@ -0,0 +1,1742 @@
...
+                    },
+                    "description": "The namespace name."
+                }, {
+                    "$ref": "#/parameters/ApiVersionParameter"
+                }, {
+                    "$ref": "#/parameters/SubscriptionIdParameter"
+                }],
+                "responses": {
+                    "200": {
+                        "description": "",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: empty or missing description not allowed. Please change this everywhere in the spec.

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 10]
```diff
@@ -0,0 +1,1742 @@
...
+                        "schema": {
+                            "$ref": "#/definitions/SharedAccessAuthorizationRuleResource"
+                        }
+                    }
+                }
+            }
+        },
+        "/{operationStatusLink}": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: why do you need this?

Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: long running operations are handled by the long running extension. Is there a reason to have this operation?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 11]
```diff
@@ -0,0 +1,1742 @@
...
+                }
+            },
+            "description": "Description of a NotificationHub AdmCredential."
+        },
+        "BaiduCredentialProperties": {
+            "properties": {
+                "baiduApiKey": {
+                    "type": "string",
+                    "description": "Get or Set Baidu Api Key."
+
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: indentation 2 spaces. Please open this in VS. Select ALL then Ctrl + K + D should format it correctly.

Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: why is NotificationHubResource not a child of "Resource". id, name, location, tags, type are properties already present in the "Resource" model. Please remove these properties from here and have an allOf on the Resource.

Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: General comment: Please remove Get or Set from the description of all the model properties. Autorest will add it automatically.

Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 12]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "string",
+                    "description": "Gets or sets the URL of the authorization token."
+                }
+            },
+            "description": "Description of a NotificationHub AdmCredential."
+        },
+        "AdmCredential": {
+            "properties": {
+                "properties": {
+                    "$ref": "#/definitions/AdmCredentialProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 13]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "string",
+                    "description": "Gets or sets the Mpns certificate Thumbprint"
+                }
+            },
+            "description": "Description of a NotificationHub MpnsCredential."
+        },
+        "MpnsCredential": {
+            "properties": {
+                "properties": {
+                    "$ref": "#/definitions/MpnsCredentialProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 14]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "string",
+                    "description": "Gets or sets the Google API key."
+                }
+            },
+            "description": "Description of a NotificationHub GcmCredential."
+        },
+        "GcmCredential": {
+            "properties": {
+                "properties": {
+                    "$ref": "#/definitions/GcmCredentialProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 15]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "string",
+                    "description": "Gets or sets the Windows Live endpoint."
+                }
+            },
+            "description": "Description of a NotificationHub WnsCredential."
+        },
+        "WnsCredential": {
+            "properties": {
+                "properties": {
+                    "$ref": "#/definitions/WnsCredentialProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 16]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "string",
+                    "description": "Gets or sets the Apns certificate Thumbprint"
+                }
+            },
+            "description": "Description of a NotificationHub ApnsCredential."
+        },
+        "ApnsCredential": {
+            "properties": {
+                "properties": {
+                    "$ref": "#/definitions/ApnsCredentialProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 17]
```diff
@@ -0,0 +1,1742 @@
...
+                "location": {
+                    "type": "string",
+                    "description": "Gets or sets Namespace data center location."
+                },
+                "name": {
+                    "type": "string",
+                    "description": "Gets or sets Name of the Namespace AuthorizationRule."
+                },
+                "properties": {
+                    "$ref": "#/definitions/SharedAccessAuthorizationRuleProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 18]
```diff
@@ -0,0 +1,1742 @@
...
+                },
+                "tags": {
+                    "type": "object",
+                    "additionalProperties": {
+                        "type": "string"
+                    },
+                    "description": "Gets or sets tags of the Namespace."
+                },
+                "properties": {
+                    "$ref": "#/definitions/NamespaceProperties",
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 19]
```diff
@@ -0,0 +1,1742 @@
...
+            },
+            "required": [
+                "location",
+                "properties"
+            ],
+            "description": "Parameters supplied to the CreateOrUpdate Namespace operation."
+        },
+        "NamespaceResource": {
+            "properties": {
+                "id": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: same comment as before. NamespaceResource must be a child of Resource. Have an allof on the resource. and remove resource properties from this model

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 20]
```diff
@@ -0,0 +1,1742 @@
...
+                    "description": "Gets or sets Namespace data center location."
+                },
+                "tags": {
+                    "type": "object",
+                    "additionalProperties": {
+                        "type": "string"
+                    },
+                    "description": "Gets or sets Namespace tags."
+                },
+                "properties": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Would you want to have `"x-ms-client-flatten": true` over here?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 21]
```diff
@@ -0,0 +1,1742 @@
...
+                "serviceBusEndpoint": {
+                    "externalDocs": {
+                        "url": "http://msdn.microsoft.com/en-us/library/windowsazure/jj873988.aspx"
+                    },
+                    "type": "string",
+                    "description": "Endpoint you can use to perform NotificationHub operations."
+                },
+                "subscriptionId": {
+                    "type": "string",
+                    "description": "The Id of the Azure subscription associated with the namespace."
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: is this the Azure subscription or the servicebus subscription?

Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: servicebus subscription and azure subscriptionId are two very different things. Hence making sure if this is the right thing.

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 22]
```diff
@@ -0,0 +1,1742 @@
...
+            "properties": {
+                "name": {
+                    "type": "string",
+                    "description": "Gets or sets name"
+                },
+                "location": {
+                    "type": "string",
+                    "description": "Gets or sets location."
+                },
+                "tags": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: Why would you want to have tags while checking the namespace availability? This seems incorrect to me.

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 23]
```diff
@@ -0,0 +1,1742 @@
...
+                    "type": "object",
+                    "additionalProperties": {
+                        "type": "string"
+                    },
+                    "description": "Gets or sets tags."
+                },
+                "isAvailiable": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: isAvailiable - spelling correction. Why would this be in the parameter? Shouldnt this be in the result?

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 24]
```diff
@@ -0,0 +1,1742 @@
...
+                    "description": "Gets or sets resource type"
+                },
+                "tags": {
+                    "type": "object",
+                    "additionalProperties": {
+                        "type": "string"
+                    },
+                    "description": "Gets or sets tags"
+                },
+                "isAvailiable": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: spelling correction - isAvailiable

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 25]
```diff
@@ -0,0 +1,1742 @@
...
+                },
+                "name": {
+                    "type": "string",
+                    "description": "Gets or sets name"
+                },
+                "type": {
+                    "type": "string",
+                    "description": "Gets or sets resource type"
+                },
+                "tags": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: please make sure if all of these properties actually make sense.
Infact having a property named "reason" which explains the reason for the namespace not being available would make more sense then the set of properties documented here.

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 26]
```diff
@@ -0,0 +1,1742 @@
...
+        }
+    },
+    "definitions": {
+        "CheckAvailabilityParameters": {
+            "properties": {
+                "name": {
+                    "type": "string",
+                    "description": "Gets or sets name"
+                },
+                "location": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: is location necessary at all? Does the service check the name availability across all the locations or a specified location?
Either the location should be required or it should be removed.

File: arm-notificationhubs/2016-03-01/swagger/notificationhubs.json [hunk 27]
```diff
@@ -0,0 +1,1742 @@
...
+                }
+            },
+            "required": [
+                "name"
+            ],
+            "description": "Parameters supplied to the Check Name Availability for Namespace and NotificationHubs."
+        },
+        "CheckAvailabilityResource": {
+            "properties": {
+                "id": {
```
Comment (amarzavery) [inline arm-notificationhubs/2016-03-01/swagger/notificationhubs.json]: the response for check availability can never be a resource. The user will be using this operation to check whether the name is available before creating a resource of that name. So how can this response have properties like id, type, location, tags. It is completely non-sensible.

Infact the response model should have a  property like reason that specifies "why something was not available" if it indeed wasn't.

Please take a look at how storage has modeled and designed a similar operation https://github.com/Azure/azure-rest-api-specs/blob/master/arm-storage/2016-01-01/swagger/storage.json#L434 and double check what your service is actually doing.

**Comment by amarzavery** [comment]
@smithab - any updates on this or shall we close this PR?

**Comment by smithab** [comment]
Sent out a changes. Wont be the fixing the CheckAvailability api yet. We are sending the Resource properties as part of its request and response. There was a CSM change in this api which we are following up on. Once we know more will update our service as well as the sdks.

**Pull request merged**